lib/graphql/schema/build_from_definition.rb

```diff
@@ -22,18 +22,60 @@ def self.call(type, field, obj, args, ctx)
 
       # @api private
       class ResolveMap
+        
         def initialize(resolve_hash)
-          @resolve_hash = resolve_hash
+          @resolve_hash = convert_keys_to_strings(resolve_hash)
+        end
+
+        def convert_keys_to_strings(h)
+          Hash[
+            h.map {|k, v|
+              v = convert_keys_to_strings(v) if v.is_a?(Hash)
+              [k.to_s, v]
+            }
+          ]
         end
 
         def call(type, field, obj, args, ctx)
           type_hash = @resolve_hash[type.name]
+
+          if !type_hash
+            type_hash = @resolve_hash[type.name] = {}
+          end
+
           type_hash && (resolver = type_hash[field.name])
 
           if resolver.nil?
-            raise(KeyError, "resolver not found for #{type.name}.#{field.name}")
-          else
-            resolver.call(obj, args, ctx)
+            raise(KeyError, "resolver not found for #{type.name}.#{field.name}") unless obj.respond_to?(field.name)
+            resolver = type_hash[field.name] = build_resolver(type, field, obj)
```

Comment:
I wonder if we could work some of this into the `@resolve_hash` itself, for example, if it was defined like this: 

```ruby
@resolve_hash = Hash.new do |h, k| 
  # For each type name, provide a new hash if one wasn't given: 
  h[k] = Hash.new do |h2, k2| 
    # For each field, provide a resolver that: 
    # - Makes a runtime check on the object then: 
    #    - if it passes, replaces itself & call through 
    #    - if it fails, raises a keyerror 
    h2[k2] = ->(obj, args, ctx) {
      if !obj.respond_to?(k2) 
        raise KeyError, "..."
      else  
        method_arity = obj.method(k2).arity 
        resolver = case method_arity
        when 0
          ->(o, a, c) { o.public_send(k2) } 
        when 1 
          ->(o, a, c) { o.public_send(k2, a) } 
        when 2 
          ->(o, a, c) { o.public_send(k2, a, c) } 
        else 
          raise "Unexpected resolve arity: #{method_arity}. Must be 0, 1, 2"
        end 
        # Call the resolver directly next time 
        h2[k2] = resolver 
        # Call through this time
        resolver.call(obj, args, ctx)
      end 
    } 
  end 
end 
```

Then, `#call` can be: 

```ruby
def call(type, field, obj, args, ctx) 
  @resolver_hash[type.name][field.name].call(obj, args, ctx)
end 
```

that keeps the default `call` behavior pretty simple and lets us use Ruby built-ins for providing fallback values in the hash. What do you think, would it work here?